### DISCUSSION TITLE
Reduce allocations due to Bindables

### DISCUSSION THREAD

**Discussion by joseph-ireland** [ppy/osu#18357] [Ideas]
I've noticed that allocations take up a disporportionate amount of time for difficulty calculation.

Today I tested a proof of concept allowing lazily creating bindables within HitObject types, and with some tweaking I am seeing that difficulty calculation runs >1.5x faster. This might also help other areas of the codebase where values can be treated as fixed, rather than editable.

To me this seems like a free performance win, and since it's possible to share a lot of the Bindable code, it doesn't necessarily add much complexity.

Is this something you might be interested in seeing developed further than just a POC? If so, any thoughts on how the type hierarchy should work? 

Code can be found here:
https://github.com/ppy/osu-framework/compare/master...joseph-ireland:IndirectBindable?expand=1 
https://github.com/ppy/osu/compare/master...joseph-ireland:IndirectBindable?expand=1

**Comment by frenzibyte** [comment]
I would focus more on why the overhead of instantiating `Bindable`s in `HitObject` affect the performance of diffcalc, as I would usually expect them to only be allocated once per beatmap, therefore never affecting it. Unless `HitObject`s are being allocated on the fly during calculation process or something of the sort?

**Reply by joseph-ireland** [reply]
Most stuff in difficulty calculation happens once per beatmap. More than half of the time is spent just constructing the beatmap, of which constructing bindables is a massive part.

**Reply by frenzibyte** [reply]
Hmm... haven't profiled, but that makes sense given the number of bindables on the massive amount of hitobjects in one beatmap.

**Comment by frenzibyte** [comment]
One alternative that doesn't involve changes to `Bindable` would be by storing a temporary field until a bindable is constructed, i.e.:
```csharp
        private Bindable<Vector2> positionBindableBacking;
        private Vector2 position;

        public Bindable<Vector2> PositionBindable => positionBindableBacking ??= new Bindable<Vector2> { Value = position };

        public virtual Vector2 Position
        {
            get => positionBindableBacking?.Value ?? position;
            set
            {
                if (positionBindableBacking != null)
                    positionBindableBacking.Value = value;
                else
                    position = value;
            }
        }
```

You can also move it to a class to avoid having `positionBindableBacking` and `position` potentially accessed across the hitobject implementation:
```csharp
    public class HitObjectProperty<T>
    {
        private Bindable<T> backingBindable;

        /// <summary>
        /// A temporary field to store the current value to, if no bindable has been constructed for this property yet.
        /// </summary>
        private T backingValue;

        public Bindable<T> Bindable => backingBindable ??= new Bindable<T> { Value = backingValue };

        public T Value
        {
            get => backingBindable != null ? backingBindable.Value : backingValue;
            set
            {
                if (backingBindable != null)
                    backingBindable.Value = value;
                else
                    backingValue = value;
            }
        }
    }
```

And wrap it in `HitObject` as such:
```csharp
        private readonly HitObjectProperty<Vector2> positionProperty = new HitObjectProperty<Vector2>();

        public Bindable<Vector2> PositionBindable => positionProperty.Bindable;

        public virtual Vector2 Position
        {
            get => positionProperty.Value;
            set => positionProperty.Value = value;
        }
```

Would like to know what the others think first before considering this approach.

**Reply by joseph-ireland** [reply]
that makes a lot of sense. I wonder how much impact using the generic would have though - it'd be an extra allocation and it'd have to box the value type too.

**Reply by frenzibyte** [reply]
I *would* hope using generic type doesn't come with too much overhead...

**Reply by joseph-ireland** [reply]
i only say that because Bindable<> is already supposed to lazily initialize the heavy weight stuff. I'll try some stuff out tomorrow, but I quite like the idea of just duplicating the value like your first example, I didn't consider doing that.

**Reply by frenzibyte** [reply]
Upon benchmarking the proposals using `BenchmarkDotNet` (which is always preferred to profile and showcase micro-optimisations) using [this simple code](https://github.com/ppy/osu/commit/8ee1722e5e98b32abd7b9aa0718290234c9211c4) (somebody double-check the code), this is what I'm seeing:

 - master (the `WithBindable` property is ineffective here):

   |    Method | Count | WithBindable |         Mean |        Error |       StdDev |    Gen 0 |    Gen 1 |   Gen 2 | Allocated |
   |---------- |------ |------------- |-------------:|-------------:|-------------:|---------:|---------:|--------:|----------:|
   | HitCircle |     1 |        False |     230.1 ns |      0.74 ns |      0.66 ns |   0.6273 |        - |       - |      1 KB |
   | HitCircle |     1 |         True |     230.9 ns |      2.75 ns |      2.44 ns |   0.6273 |        - |       - |      1 KB |
   | HitCircle |   100 |        False |  26,799.0 ns |    532.76 ns |    946.98 ns |  49.5911 |  16.5100 |       - |    126 KB |
   | HitCircle |   100 |         True |  27,429.2 ns |    460.03 ns |    384.15 ns |  49.5300 |  16.5100 |       - |    126 KB |
   | HitCircle |  1000 |        False | 832,436.4 ns | 16,642.34 ns | 17,807.12 ns | 590.8203 | 289.0625 | 80.0781 |  1,258 KB |
   | HitCircle |  1000 |         True | 808,209.9 ns |  3,255.44 ns |  2,885.86 ns | 590.8203 | 289.0625 | 78.1250 |  1,258 KB |

 - [lazy-init bindable / `HitObjectProperty` approach](https://github.com/ppy/osu/compare/d84119ac0631367b5b3c0f60a9b56c4aa66e900c..31ad47a190e6526c05f57f73780df103d4ca9855) (open for better name):

   |    Method | Count | WithBindable |           Mean |        Error |       StdDev |    Gen 0 |    Gen 1 |    Gen 2 |   Allocated |
   |---------- |------ |------------- |---------------:|-------------:|-------------:|---------:|---------:|---------:|------------:|
   | HitCircle |     1 |        False |       134.5 ns |      1.20 ns |      1.12 ns |   0.3481 |        - |        - |       728 B |
   | HitCircle |     1 |         True |       269.8 ns |      3.33 ns |      3.12 ns |   0.7534 |        - |        - |     1,576 B |
   | HitCircle |   100 |        False |    15,429.5 ns |    295.19 ns |    303.14 ns |  33.6609 |   2.6398 |        - |    70,424 B |
   | HitCircle |   100 |         True |    36,810.4 ns |    735.03 ns |    902.68 ns |  66.5894 |  27.8931 |        - |   155,224 B |
   | HitCircle |  1000 |        False |   351,410.4 ns |  6,944.67 ns | 10,394.46 ns | 335.9375 | 167.9688 |        - |   704,024 B |
   | HitCircle |  1000 |         True | 1,076,059.5 ns | 20,985.35 ns | 27,286.89 ns | 628.9063 | 287.1094 | 113.2813 | 1,552,101 B |

 - [`IndirectBindable` approach](https://github.com/ppy/osu/compare/d84119ac0631367b5b3c0f60a9b56c4aa66e900c..aff174643dbe555d505cbdea408ce30f4bcd21a7):
   
   |    Method | Count | WithBindable |           Mean |        Error |       StdDev |    Gen 0 |    Gen 1 |    Gen 2 |   Allocated |
   |---------- |------ |------------- |---------------:|-------------:|-------------:|---------:|---------:|---------:|------------:|
   | HitCircle |     1 |        False |       122.0 ns |      2.46 ns |      3.75 ns |   0.3328 |        - |        - |       696 B |
   | HitCircle |     1 |         True |       344.6 ns |      4.87 ns |      4.07 ns |   1.2240 |        - |        - |     2,560 B |
   | HitCircle |   100 |        False |    12,622.2 ns |    230.13 ns |    192.17 ns |  31.0516 |   4.2114 |        - |    67,224 B |
   | HitCircle |   100 |         True |    68,150.9 ns |  1,044.75 ns |    926.14 ns | 120.7275 |  60.3027 |        - |   253,624 B |
   | HitCircle |  1000 |        False |   314,309.9 ns |  5,850.25 ns |  5,186.09 ns | 320.3125 | 160.1563 |        - |   672,024 B |
   | HitCircle |  1000 |         True | 1,727,894.1 ns | 34,313.75 ns | 40,848.08 ns | 789.0625 | 318.3594 | 115.2344 | 2,536,103 B |

Safe to say there's only a negligible difference between the lazy-init bindable approach and the `IndirectBindable` approach, don't see much of an impact from the use of generics.

**Reply by joseph-ireland** [reply]
I think if you make HitObjectProperty a struct, you'll match master performance for WithBindable and IndirectBindable for WithBindable=false. I know there's only 10% difference or something, but i guess there's no reason not to

**Reply by joseph-ireland** [reply]
i.e. here's with class HitObjectProperty:

|    Method | Count | WithBindable |          Mean |        Error |       StdDev |    Gen 0 |    Gen 1 |   Allocated |
|---------- |------ |------------- |--------------:|-------------:|-------------:|---------:|---------:|------------:|
| HitCircle |     1 |        False |      96.54 ns |     0.435 ns |     0.386 ns |   0.1740 |        - |       728 B |
| HitCircle |     1 |         True |     207.98 ns |     0.853 ns |     0.756 ns |   0.3767 |        - |     1,576 B |
| HitCircle |   100 |        False |   9,882.38 ns |    32.516 ns |    27.152 ns |  16.8304 |        - |    70,424 B |
| HitCircle |   100 |         True |  27,326.20 ns |   318.046 ns |   297.501 ns |  32.1045 |  11.3831 |   155,224 B |
| HitCircle |  1000 |        False | 183,285.79 ns | 1,005.944 ns |   940.961 ns | 112.0605 |  55.9082 |   704,024 B |
| HitCircle |  1000 |         True | 526,334.47 ns | 4,861.616 ns | 4,547.558 ns | 247.0703 | 123.0469 | 1,552,024 B |

vs struct HitobjectProperty https://github.com/ppy/osu/compare/master...joseph-ireland:hit_object_property_struct?expand=1

|    Method | Count | WithBindable |          Mean |        Error |       StdDev |    Gen 0 |    Gen 1 |   Allocated |
|---------- |------ |------------- |--------------:|-------------:|-------------:|---------:|---------:|------------:|
| HitCircle |     1 |        False |      55.70 ns |     1.148 ns |     1.074 ns |   0.1243 |        - |       520 B |
| HitCircle |     1 |         True |     173.17 ns |     1.463 ns |     1.297 ns |   0.3271 |        - |     1,368 B |
| HitCircle |   100 |        False |   6,003.48 ns |    51.256 ns |    47.945 ns |  11.8637 |   0.0076 |    49,624 B |
| HitCircle |   100 |         True |  23,247.37 ns |   219.320 ns |   205.152 ns |  28.8086 |   9.8572 |   134,424 B |
| HitCircle |  1000 |        False |  93,863.66 ns |   462.840 ns |   386.492 ns |  78.9795 |  39.4287 |   496,024 B |
| HitCircle |  1000 |         True | 446,401.23 ns | 2,958.636 ns | 2,470.595 ns | 213.8672 | 106.9336 | 1,344,024 B |

**Comment by joseph-ireland** [comment]
I guess one downside in any lazy init approach is the potential for race conditions, or locking overhead if that's dealt with properly

**Reply by frenzibyte** [reply]
I'm not sure there'll *ever* be a case where a `HitObject` is accessed across multiple threads, but if that potentially happens, then using lock seems fine I think.

**Comment by joseph-ireland** [comment]
made a POC for both alternative implementations: 
https://github.com/ppy/osu/compare/master...joseph-ireland:manual_lazy_bindable?expand=1
https://github.com/ppy/osu/compare/master...joseph-ireland:lazy_bindable_hitobjects?expand=1

If you make the generic a struct then there is no overhead, both perform the same, so it seems using a `struct LazyBindable<T>` is the best option.

**Reply by joseph-ireland** [reply]
I guess there are additional considerations re default values, bounds, etc, and specializations like BindableDouble

**Comment by peppy** [comment]
We've already discussed internally the potential of removing bindables completely from `HitObject` for the same reasoning. A bit of a longer term goal / consideration.

**Reply by joseph-ireland** [reply]
Do you have a plan on how to achieve that? I guess you could make a BindableHitObject type or something for use in the editor. Any approach is going to need a pretty big refactor though

**Reply by joseph-ireland** [reply]
Would have been easier if it was a shorter term goal before significant work went into the editor

**Reply by peppy** [reply]
It's not of priority to us currently so I do not have an immediate answer. Difficulty calculation is already of ample performance compared to other work that needs to be done.

**Reply by joseph-ireland** [reply]
To clarify: this isn't the direction you anticipated, and you're not interested in these changes, even as a short term fix?

**Reply by peppy** [reply]
Lazy init may be fine, if there's a valid use case for it. You haven't specified why the allocations are an issue in the first place, so context would be required.

**Reply by joseph-ireland** [reply]
Profiling shows a bunch of time spent in GC and creating bindables, and changing the implementation gives significant performance improvements. 

I have a beatmap test folder which calculates in 11.8 seconds on master, 10.5 seconds in these test branches above (mininal/trivial changes only in OsuHitObject), and 8.1 seconds in the original branch with more extensive changes (should also be achievable using a LazyBindable<T>).

Would you want any other context?

**Comment by smoogipoo** [comment]
I'm weary of overcomplicating bindables by adding indirection like this, since it seems like a sticky tape fix when bindables are likely the wrong construct for this in the first place. But not only that, this inefficiency is probably exhibited in other areas of the game as well, so are we looking for more of a systemic change or a local temporary thing?

For a systemic change, maybe this is something for o!f to handle instead via some `BindableSlim<>` struct on the stack, for example:
https://github.com/smoogipoo/osu-framework/tree/bindable-slim
https://github.com/smoogipoo/osu/tree/bindable-slim
It's similar to the `HitObjectProperty<>` variant from the above, except that it proxies every `Bindable<>` property for end-user ease of use - it behaves exactly like `Bindable<>`.

For a temporary change, the `IndirectBindable` variant is unsafe since it separates `Default`/`Disabled` states - what happens if `PositionBindable` is disabled but the user then sets a value via `Position`? I definitely prefer the `HitObjectProperty<>` variant from above from this perspective, as it should handle things safely as far as I can tell.

I think I'd be fine with getting in the `HitObjectProperty<>` variant as a temporary change and seeing how that evolves.

**Comment by frenzibyte** [comment]
Lazy-initialisation has been added at https://github.com/ppy/osu/pull/19225.
